### ISSUE TITLE
Performance question

### ISSUE THREAD

**Issue by EelcoHoogendoorn** [brentyi/jaxlie#1]
Hi, thanks for creating what looks to be a very useful library.

Not an issue here but just a question; what has your experience been in terms of performance, working with JAX in this style? Of course by implementing this in JAX we get all the nice autodiff benefits which I am also interested in. 

But does the expectation bear out, that operating on a single quaternion in this style, is of comparable slowness as working in plain python or numpy? That is, dominated by the overhead of invoking the individual instructions? Ofcourse JAX gives us automatic vectorization as well; but do you know how that works out in practice? Do we trend towards vectorized-numpy type performance on the CPU, if acting on batches of objects? And how large batches are we talking about, before more or less constant time complexity goes to linear time complexity?

Of course these are simple questions I can benchmark myself, but my question is more an open one about your experiences pertaining to performance, working with JAX in this style. Thanks in advance!

**Comment by EelcoHoogendoorn** [comment]
```
from jaxlie import SE3, SO2, SO3

import jax
from jax import vmap, make_jaxpr
import jax.numpy as jnp
import jax.random

key = jax.random.PRNGKey(0)

@jax.jit
def bar(arr: jnp.array):
	q = SO3(arr)
	return q.as_matrix()

m = vmap(bar)
I = jnp.identity(4)[:2]
print(m(I))

import timeit
I = jnp.identity(4)[0]
print(timeit.timeit(lambda : bar(I), number=1000))

for n in range(6):
	I = jax.random.normal(key, (10**n, 4))
	print(n, timeit.timeit(lambda : m(I), number=1000))

```

Potentially interesting; it kinda surprises me:

```
0.14312348899999994
0 0.43089010400000016
1 0.4277055970000001
2 0.4243873729999996
3 0.43083932199999975
4 0.43889132099999983
5 1.4538018470000003
```

So, vmap adds some overhead, about 3x in this case. That kinda surprises me since the overhead of invoking the JAX function is indeed massively dominant; we need a batch of a 100k rows to start seeing anything but overhead in our benchmark.


To get some idea if this is overhead of calling into JAX, or JAX invoking individual array operations, we can doubling the work done...
```
@jax.jit
def bar2(arr: jnp.array):
	q = SO3(arr)
	q2 = SO3(arr*2)

	return q.as_matrix(), q2.as_matrix()
m = vmap(bar2)

for n in range(6):
	I = jax.random.normal(key, (10**n, 4))
	print(n, timeit.timeit(lambda : m(I), number=1000))
```
... which has the following effect:
```
0 0.610528499
1 0.5946009659999998
2 0.578929413
3 0.5877716949999989
4 0.5984280680000005
5 2.9029480670000005
```
Which suggest both play a role I suppose. At 100k rows, we get something pretty close to a clean doubling, but at the low end of batch size, it is implied the original bar runtime is about 50/50 calling-into-JAX vs calls-within-JAX. So we definitely do want to JIT our operations together as much as possible, but regardless, the conclusion that we need batches of a 100k not to be overhead-bound does not drastically change, sadly.

Note that this is all for a CPU backend; the ratios will undoubtly be worse on an accelerator. I was hoping for a little better than this, although I suppose the as_matrix function with all its scalar operations is a worst-case, and inspecting the JAX expression makes it look even worse (tons of concatting going on there; does not seem like it allocates the fixed-size 4x4 and then modify in place; unless these type of optizations happen in a later stage? could imagine they dont given the novelty of the project). If there are any opportunities to perform vectorized operations on smaller internal axes, that should be a huge help.

I intend to write a little differentiable physics simulator in JAX, and I do want to do my rollouts in batches... but probably not of 100k, so I better keep my eyes open for additional vectorization opportunities.

**Comment by EelcoHoogendoorn** [comment]
Noticed I needed another jit after vmap-ing a jitted function; that makes the vmap-ed and non-vectorized timings pretty close, so eliminating that mystery. Now you can see some impact of N at 10k; but still you need at least 100k not to be very overhead-dominated.

**Comment by brentyi** [comment]
Thanks for writing these benchmarks @EelcoHoogendoorn!

Are you intentionally including the JIT compile time into your method timing? Note that this JIT compilation needs to happen once for each unique input shape.

I'd suggest one more change:

```python
from jaxlie import SE3, SO2, SO3

import jax
from jax import vmap, make_jaxpr
import jax.numpy as jnp
import jax.random

key = jax.random.PRNGKey(0)

@jax.jit
def bar(arr: jnp.array):
    q = SO3(arr)
    return q.as_matrix()

m = jax.jit(vmap(bar))  # New: JIT around the vmapped function
I = jnp.identity(4)[:2]
print(m(I))

import timeit
I = jnp.identity(4)[0]
print(timeit.timeit(lambda: bar(I), number=1000))

for n in range(6):
    I = jax.random.normal(key, (10 ** n, 4))
    m(I)  # New: don't time the first call, whose runtime will be dominated by tracing
    print(n, timeit.timeit(lambda: m(I), number=1000))
```
This outputs:
```
0.09984651500417385
0 0.0021195340013946407
1 0.0020842389931203797
2 0.0031986079993657768
3 0.011603418002778199
4 0.08440414799406426
5 0.4854047819972038
```

and, when the workload is doubled:
```
0 0.0026417010012664832
1 0.0027852979983435944
2 0.004813161001948174
3 0.021309905998350587
4 0.38269638800557004
5 2.7484699150008964
```

**Comment by brentyi** [comment]
Thanks for writing these benchmarks @EelcoHoogendoorn!

Are you intentionally including the JIT compile time into your method timing? Note that JIT compilation will happen once for each unique input shape.

I'd suggest one more change:

```python
from jaxlie import SE3, SO2, SO3

import jax
from jax import vmap, make_jaxpr
import jax.numpy as jnp
import jax.random

key = jax.random.PRNGKey(0)

@jax.jit
def bar(arr: jnp.array):
    q = SO3(arr)
    return q.as_matrix()

m = jax.jit(vmap(bar))  # New: JIT around the vmapped function
I = jnp.identity(4)[:2]
print(m(I))

import timeit
I = jnp.identity(4)[0]
print(timeit.timeit(lambda: bar(I), number=1000))

for n in range(6):
    I = jax.random.normal(key, (10 ** n, 4))
    m(I)  # New: don't time the first call, whose runtime will be dominated by tracing
    print(n, timeit.timeit(lambda: m(I), number=1000))
```
This outputs:
```
0.09984651500417385
0 0.0021195340013946407
1 0.0020842389931203797
2 0.0031986079993657768
3 0.011603418002778199
4 0.08440414799406426
5 0.4854047819972038
```

and, when the workload is doubled:
```
0 0.0026417010012664832
1 0.0027852979983435944
2 0.004813161001948174
3 0.021309905998350587
4 0.38269638800557004
5 2.7484699150008964
```

**Comment by EelcoHoogendoorn** [comment]
> Note that JIT compilation will happen once for each unique input shape.

Ah! No I did not know that, I was assuming it was cached on type only like numba. (spot the JAX noob)

That is a lot more reasonable; seems we are getting pretty linear scaling around the 1k rows mark now. Batches of 1k are probably ballpark reasonable for my application; plus the totality of the code will have more vectorization opportunities than these particular operations which are pretty worst-case.

On my machine im hitting linear scaling at 10k now, with timings pretty flat below that; plus my shortest runtime is still 10x yours. (±2e-2 compared to your ±2e-3). What hardware are you running on? This is a 2019MBP; not the fastest out there but im not aware of any CPUs that should outperform it 10x single-core; and as I understand there is no automatic multithreading; nor would I expect the unvectorized case to benefit from it. 

Are there any other factors you know to be at play here?

**Comment by brentyi** [comment]
Okay great!

Sounds promising, though the performance delta you're seeing is certainly unexpected. I'm not running on any crazy 10x hardware, just my laptop:
- Lenovo X1 Yoga 4, i7-10710u
- Ubuntu 20.04
- Python 3.8.8 (also checked in 3.7)
- `jax==0.2.16`, `jaxlib==0.1.68` 

Can you see if you're on an older version of JAX? Upgrading might help.

**Comment by brentyi** [comment]
Downgrading to `jaxlib==0.1.67` results in numbers similar to what you're describing:
```
0 0.014290804996562656
1 0.013791656005196273
2 0.01471616299386369
3 0.013255646998004522
4 0.13410871299856808
5 1.265380417004053
```

**Comment by EelcoHoogendoorn** [comment]
Yeah I was on OSX py38, jax 0.2.16, jaxlib 0.1.67; I went with the conda install which doesnt have the newer jaxlib yet. Upgrading to 0.1.68 does indeed give a comparable pattern to what you see.

```
0 0.003444935999999954
1 0.0032514040000000133
2 0.004201270999999895
3 0.01406876999999973
4 0.10828858600000002
5 0.5346468250000003
```
```
0 0.004277849000000167
1 0.0042401299999994535
2 0.005829367999999668
3 0.023577491000000173
4 0.18599050100000003
5 1.77556153
```

Interestingly the highest-workload case is a fair bit faster for me whereas the low workload cases are still almost 2x slower, but im willing to chalk that up to to OS/CPU/cache/mem differences, despite them both being fairly recent mobile i7s.

Great to see JAX is still improving at such a fast pace. Tried to find a changelog to get a sense of what happened here but it seems they are not super consistent in updating that.
